### PR TITLE
[Program:GCI] Add user name search functionality in GET /users API

### LINKED ISSUE #389: Add query parameter to search users by user display name to GET /users API
Proposed by @annabauza during GCI contest

## Description
As a user of mentorship system,
I need to be able to search users by name,
so that I can look for someone I'm interested in having a mentorship relation with.

GET /users API needs a way to search users by user display name. This should be done by adding `search=`parameter to url. Value of this parameter should be decode from url encoding.

### Update [Required]
- [ ] Add a query parameter to extend `GET /users` by search parameter `search=[search_query]`
- [ ] Return users which `display user name` match search query 
- [ ] Update Swagger UI documentation for this API
- [ ] Write at least one test to prove your new feature is working

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
4 hours

### PR DESCRIPTION
**Pull request by daksha-aeer**
### Description
User name search functionality added to GET /users endpoint. Term to search for is taken through query string parameters.

Fixes #389 

#### Working:
1. New query string parameter 'search' added for GET /users route. If 'search' is empty then show all users, otherwise search for the query.
2. **'flask_whooshalchemyplus'** module installed for search functionality. It creates search index for fields specified in '**\_\_searchable__**'. Woosh index created for **'name'** and **'username'**.
3. **search_users()** function created in dao/user.py. It searches for the specified query in database. This function is called by GET /users route.

#### References:
https://www.youtube.com/watch?v=bAPcmGNsulc
https://github.com/Revolution1/Flask-WhooshAlchemyPlus

### Type of Change:
- Code

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. Swagger description
task65_1

2. Searching for term
task65_2

3. Unit tests
```sh
python -m unittest discover tests
```
task65_3

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@daksha-aeer I haven't used this whoosh extension before so i learned something here.
Thank you so much for your contribution 🎉 Could you add a simple test, showing when you query something it returns the results intended?

**Comment by daksha-aeer** [comment]
> @daksha-aeer I haven't used this whoosh extension before so i learned something here.
> Thank you so much for your contribution  Could you add a simple test, showing when you query something it returns the results intended?

Done, please review : )

File: tests/users/test_api_list_users.py [hunk 1]
```diff
@@ -50,6 +50,19 @@ def test_list_users_api_resource_auth(self):
         self.assertEqual(200, actual_response.status_code)
         self.assertEqual(expected_response, json.loads(actual_response.data))
 
+    def test_search_user_api_resource_auth(self):
+        auth_header = get_test_request_header(self.admin_user.id)
+        print("verified user: %s"%self.verified_user.name)
+        print("other user: %s"%self.other_user.name)
+        expected_response = marshal(self.verified_user, public_user_api_model)
+        actual_response = self.client.get(
+            '/users',data = {'search': self.verified_user.name},
```
Comment (annabauza) [inline tests/users/test_api_list_users.py:57]: Can we search partial as well?

File: tests/users/test_api_list_users.py [hunk 2]
```diff
@@ -50,6 +50,19 @@ def test_list_users_api_resource_auth(self):
         self.assertEqual(200, actual_response.status_code)
         self.assertEqual(expected_response, json.loads(actual_response.data))
 
+    def test_search_user_api_resource_auth(self):
+        auth_header = get_test_request_header(self.admin_user.id)
+        print("verified user: %s"%self.verified_user.name)
```
Comment (isabelcosta) [inline tests/users/test_api_list_users.py]: Can you remove the print statements? I'm sure you used them just for testing, right?

**Review by annabauza — CHANGES_REQUESTED** [review]

File: tests/users/test_api_list_users.py [hunk 3]
```diff
@@ -50,6 +50,19 @@ def test_list_users_api_resource_auth(self):
...
+        print("other user: %s"%self.other_user.name)
```
Comment (isabelcosta) [inline tests/users/test_api_list_users.py]: this one too?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
There are two unnecessary print statements, can you please remove them?

File: tests/users/test_api_list_users.py (hunk 1 above)
Comment (daksha-aeer) [reply tests/users/test_api_list_users.py:57]: Yes Anna. In the attached screenshot I searched for term "gci". This gave various names containing the term like gcitaskthree, gciuserfive and so on.

File: tests/users/test_api_list_users.py (hunk 3 above)
Comment (daksha-aeer) [reply tests/users/test_api_list_users.py]: Done!

File: tests/users/test_api_list_users.py (hunk 2 above)
Comment (daksha-aeer) [reply tests/users/test_api_list_users.py]: done!

**Review by isabelcosta — APPROVED** [review]
Thank you for adding a test @daksha-aeer :)

**Comment by isabelcosta** [comment]
@ daksha-aeer the task is complete, but TravisCI is failing https://travis-ci.org/systers/mentorship-backend/builds/637669631 we have to see this before merging the PR. cc @annabauza

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@daksha-aeer  can you please see why the TravisCI is failing and fix it.

**Comment by daksha-aeer** [comment]
> @daksha-aeer can you please see why the TravisCI is failing and fix it.

Travis issue fixed!

**Review by annabauza — APPROVED** [review]

**Comment by isabelcosta** [comment]
@annabauza @daksha-aeer this commit still causes TravisCI to fail https://travis-ci.org/systers/mentorship-backend/builds/644877926 :/

**Pull request merged**